src/System.Management.Automation/engine/pipeline.cs

```diff
@@ -220,44 +220,59 @@ private void Log(string logElement, InvocationInfo invocation, PipelineExecution
                 }
             }
 
-            if (!string.IsNullOrEmpty(logElement))
+            if (NeedToLog() && !string.IsNullOrEmpty(logElement))
```

Comment:
1. I guess original design is for speed. 
2. The design was:
2.1 _Always_ collect events
2.2. Write all collected events to a log if logging is enabled at pipeline end time.
3. New behavior:
3.1 _Skip_ events until logging is enabled
3.2 Write all collected events to a log if logging is enabled at pipeline end time.

Can you conform new behavior is well?

If so new design allows enable logging while the pipeline run but we call NeedToLog() that is expensive. Can we avoid this performance regression?

Reply:
Can we maybe check for this at pipeline creation and store it as a field on the pipeline processor to reference here instead?

How often would we expect this value to change during the execution of a pipeline? Does it necessarily need to take effect for already-running pipelines?

Reply:
> Does it necessarily need to take effect for already-running pipelines?

It could make sense for long running pipelines (specially if this helps to diagnostic "hang" scenarios).

Reply:
Whether we need to log an event or not is determined when finishing add commands to a pipeline, and it cannot be changed during the pipeline execution.

`NeedToLog` returns `true` when any command processor in the pipeline has `Command.commandRuntime.LogPipelineExecutionDetail` evaluated to `true`.

The `LogPipelineExecutionDetail` is a get-only property. Its value is set in `MshCommandRuntime` constructor, and is determined by `PSSnapIn.LogPipelineExecutionDetails` or `Module.LogPipelineExecutionDetails`.

So, even though `PSSnapIn.LogPipelineExecutionDetails` and `Module.LogPipelineExecutionDetails` has public setter/getter and a user can change their values, the `MshCommandRuntime.LogPipelineExecutionDetail` property won't change once the `MshCommandRuntime` instance is constructed. The `MshCommandRuntime` is constructed when constructing a command processor, be it `CommandProcessor`, `ScirptCommandProcessor`, or `NativeCommandProcessor`. After commands are added to a `PipelineProcessor`, it's already determined whether we will write out a pipeline execution-detail event in the end. This won't change during the pipeline execution.

The old behavior is: always collecting all logs, even if we will never write the pipeline execution-detail log in the end;
The new behavior is: only collecting logs when we are going to write out the event.

Reply:
> Can we maybe check for this at pipeline creation and store it as a field on the pipeline processor to reference here instead?

The new change already uses a field for that purpose, and the field is set when `NeedToLog()` gets called for the first time. The reason to still keep `NeedToLog()` method is that `LogToEventLog` is called in `DisposeCommands()`, which may be called more than once (e.g. in a steppable pipeline scenario). So the `if (_commands is null) { return false }` check needs to be kept.

Reply:
@daxian-dbw Thanks for clarify!

The implementation was complicated and now it's even more confusing.
I suggest to move the logic from NeedToLog() to Start() method. If we make _eventLogBuffer initialization in Start() we have no need _needToLog at all. This will accurately reflect the design intent you have described.
Here the code will be as simple as:
```c#
            if (_eventLogBuffer is not null && !string.IsNullOrEmpty(logElement))
            {
                _eventLogBuffer.Add(logElement);
            }
```

Reply:
I don't think the `NeedToLog` logic is anywhere near "complicated" 😄, but I agree it can be simplified even further.
I pushed a change to set `_needToLog` in `PipelineProcessor.AddCommand`, so the `_needToLog` will be determined after all commands are added for a pipeline.

I insist to keep `_eventLogBuffer ??= new List<string>();` instead of initializing the buffer somewhere in `AddCommand` or `Start`. You may notice the comment in `LogToEventLog()`:
> // We check to see if there is anything in the buffer before we flush it.
> // Flushing the empty buffer causes a measurable performance degradation.

That means there are cases that we need to log, but there is actually no log message written to the buffer. In such a case, it's better to not even allocate the buffer list.

Reply:
> I don't think the NeedToLog logic is anywhere near "complicated"

I mean how the logging works in this place as a whole. It would be more useful to be able to enable logging at any time. Since this is not possible, it makes no sense to do any dynamic checks and initializations but simplify code so that reflect to the design exactly.

> it's better to not even allocate the buffer list.

But now we allocate nullable bool.
The List allocation is about some bytes per pipeline. It is not critical since Pipeline allocates ~10 objects.

Reply:
> But now we allocate nullable bool.

Nope, it's not a nullable bool anymore.

> It is not critical since Pipeline allocates ~10 objects.

Why allocate the list when it's so simple to avoid it?